File: Samples/sample.ts
type Metric = {
  name: string;
  value: number;
  tags: Record<string, string>;
};

const metrics: Metric[] = [
  { name: "latency.p95", value: 184.2, tags: { env: "staging" } },
  { name: "errors", value: 2, tags: { service: "api" } },
];

function formatMetric(metric: Metric): string {
  const tags = Object.entries(metric.tags).map(([key, value]) => `${key}=${value}`).join(",");
  return `${metric.name.padEnd(16)} ${metric.value.toFixed(2)} ${tags}`;
}

metrics.map(formatMetric).forEach((line) => console.log(line));

File: Samples/sample.rs
#[derive(Debug, Clone)]
struct Metric {
    name: String,
    value: f64,
    tags: Vec<(String, String)>,
}

impl Metric {
    fn new(name: &str, value: f64, tags: Vec<(&str, &str)>) -> Self {
        Self {
            name: name.to_string(),
            value,
            tags: tags.into_iter().map(|(k, v)| (k.to_string(), v.to_string())).collect(),
        }
    }

    fn render(&self) -> String {
        let tags = self.tags.iter().map(|(k, v)| format!("{k}={v}")).collect::<Vec<_>>().join(",");
        format!("{:<16} {:>8.2} {}", self.name, self.value, tags)
    }
}

fn main() {
    let metrics = vec![
        Metric::new("latency.p95", 184.2, vec![("env", "staging")]),
        Metric::new("errors", 2.0, vec![("service", "api")]),
    ];
    for metric in metrics {
        println!("{}", metric.render());
    }
}
